### PR TITLE
Improve sharding and dispatch

### LINKED ISSUE #104922: anti-windup mechanism in queueset is too aggressive
#### What happened:
I looked into why TestDifferentWidth in `queueset_test.go` needs surprisingly wide tolerance for divergence from the ideal result.  One reason is the anti-windup mechanism at https://github.com/kubernetes/kubernetes/blob/b47f8263e18c7b13dba33fba23187e5e0477cdbd/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L726-L742 is canceling the "credit" that the queues for the width=1 flow get from having requests wait in the queue.  For example, here is a bit of verbose logging from around the time the first round of requests finish and the second start (I am using a tweaked scenario, hoping a wider one will come closer to ideal --- it did not):
```
I0910 17:15:33.397998   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=70 finishR=0.006000064296031746
I0910 17:15:33.398044   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=1 finishR=1.003
I0910 17:15:33.398084   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=2 finishR=2.006000002
I0910 17:15:33.398127   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=5 finishR=0.0030000732960317462
I0910 17:15:33.398167   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=10 finishR=0.0090000155
I0910 17:15:33.398209   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=19 finishR=0.012000018166666668
I0910 17:15:33.398247   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=29 finishR=0.006000050823809524
I0910 17:15:33.398284   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=33 finishR=0.012000034466666667
I0910 17:15:33.398327   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=35 finishR=0.003000083296031746
I0910 17:15:33.398368   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=49 finishR=0.0060000479666666665
I0910 17:15:33.398410   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=52 finishR=0.009000031666666667
I0910 17:15:33.398449   39294 queueset.go:697] "Considering queue to dispatch" queueSet="TestDifferentWidths" queue=61 finishR=0.8393333616293651
I0910 17:15:33.398499   39294 queueset.go:641] QS(TestDifferentWidths) at r=2021-09-10 17:15:34.357788002 v=0.833333363ss: dispatching request "TestDifferentWidths" []int{0, 5, 0} work {1 0s} from queue 5 with start R 0.833333363ss, queue will have 5 waiting & 1 requests occupying 1 seats, set will have 9 seats occupied
```

The "considering" line showed that queue 5 had quite a small value of `virtualStart`, but the "dispatching" line shows that it has jumped by about 0.83ss (which is most of the roughly 1ss credit it got from sitting out the first round).

#### Environment:
- Kubernetes version (use `kubectl version`): commit 59d319ec06bb33289a87036418b4a61ed3bb215f with TestDifferentWidths tweaked to have more concurrency
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR makes improvements to queueset shuffle sharding and dispatching as well as further improvements to the queueset tests.

- Switch to a new anti-windup technique because the old one was too aggressive.  The new one's floor on a queue's next dispatch time is the arrival time of the request to be dispatched.
- When judging queues in the shuffle-sharding, look at seats in use as well as seats of waiting requests, because this is a more accurate reflection of how busy the queue is.
- In shuffle-sharding, do not always consider the items in the hand in the same order.  Before this change, I found that in scenarios with multiple flows going to the same queue, the flows got uneven treatment depending on the queue's position in each flow's hand.

With the above changes, we get better results.  This PR tightens the margins of several tests to reflect this.  This PR also makes some improvements in comments and private method names.

#### Which issue(s) this PR fixes:

Fixes #104922 

#### Special notes for your reviewer:
This PR addresses more than #104922, but that is the only formal issue.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-misc 
/cc @wojtek-t

**Comment by MikeSpreitzer** [comment]
/cc @tkashem 
/cc @deads2k 
@lavalamp

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 1]
```diff
@@ -471,23 +475,33 @@ func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(ctx context.Conte
 // chooseQueueIndexLocked uses shuffle sharding to select a queue index
 // using the given hashValue and the shuffle sharding parameters of the queueSet.
 func (qs *queueSet) chooseQueueIndexLocked(hashValue uint64, descr1, descr2 interface{}) int {
+	var backHand [8]int
+	// the dealer uses the current desired number of queues, which is no larger than the number in `qs.queues`.
+	hand := qs.dealer.DealIntoHand(hashValue, backHand[:])
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:480]: Can you describe what's the purpose of this change?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 2]
```diff
@@ -723,22 +738,21 @@ func (qs *queueSet) selectQueueLocked() *queue {
...
+
+	// Next comes the anti-windup hack.
+	// We need a hack because all non-empty queues are allocated the same
+	// number of seats.  A queue that can not use all those seats and does
+	// not go empty accumulates a progresively earlier `virtualStart` compared
+	// to queues that are using more than they are allocated.
+	// The following hack addresses the first side of that inequity,
+	// by insisting that dispatch in the virtual world not precede arrival.
+	var virtualStartBound = oldestReqFromMinQueue.arrivalR
+	if minQueue.virtualStart < virtualStartBound {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:750]: Thinking about this - I can imagine that after adjusting the `minQueue` is no longer the queue that we should choose.

Shouldn't we actually do those adjustments above (when choosing the queue in line 701-719)?

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:750]: Better yet, modify the code that currently introduces the suppressed situation.

**Comment by MikeSpreitzer** [comment]
The force-push to 9a7e054e05e is a rebase onto `master` and gives names to test cases in TestNoRestraint and TestWindup.

**Comment by MikeSpreitzer** [comment]
The force-push to ca4b609c9a4 responds to the above review comments and improves the names of the methods that do the shuffle sharding and pick the queue to dispatch from.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:483]: done.

**Comment by MikeSpreitzer** [comment]
The force-push to 4b9cba85874 removes one outdated comment (that reflected the earlier anti-windup technique).

**Comment by MikeSpreitzer** [comment]
Following is a rendering of a run of TestTooWide.

[toowide.pdf](https://github.com/kubernetes/kubernetes/files/7208060/toowide.pdf)

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 3]
```diff
@@ -468,26 +474,37 @@ func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(ctx context.Conte
...
+		waitingSeats := queue.requests.SeatsSum()
 		// TODO: Consider taking into account `additional latency` of requests
 		// in addition to their seats.
 		// Ideally, this should be based on projected completion time in the
 		// virtual world of the youngest request in the queue.
-		queue := qs.queues[queueIdx]
-		waitingSeats := queue.requests.SeatsSum()
-		thisSeatsSum := waitingSeats // + queue.seatsInUse
-		klog.V(7).Infof("QS(%s): For request %#+v %#+v considering queue %d of %d seats waiting and %d executing", qs.qCfg.Name, descr1, descr2, queueIdx, waitingSeats, queue.seatsInUse)
+		thisSeatsSum := waitingSeats + queue.seatsInUse
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:497]: @tkashem - I think you commented out seatsInUse in the past, but I can't remember the exact reason. Can you remind this?

**Review by wojtek-t — COMMENTED** [review]
Just one comment to remind me why did we commented the seats-waiting before.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 3 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:497]: It was giving worse results.
With the other changes, it now gives good enough results.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:497]: this line was added by Mike's PR https://github.com/kubernetes/kubernetes/pull/104345/files#diff-9411bfa852eec5372d491bd393f19ef57cab8d5cd54af3a842754dcc327f68a7R482-R484. 
```
queue := qs.queues[queueIdx]
waitingSeats := queue.requests.SeatsSum()
thisSeatsSum := waitingSeats // + queue.seatsInUse
```

I believe un-commenting `queue.seatsInUse` makes sense here.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:497]: SGTM
